baremetal/machine_manager_test.go

```diff
@@ -0,0 +1,1467 @@
+	}
+
+	addr3 := capi.MachineAddress{
+		Type:    capi.MachineHostName,
+		Address: "mygreathost",
+	}
+
+	testCases := map[string]struct {
+		Machine               capi.Machine
+		BMMachine             capbm.BareMetalMachine
+		Host                  *bmh.BareMetalHost
+		ExpectedNodeAddresses []capi.MachineAddress
+	}{
+		"One NIC": {
+			Host: &bmh.BareMetalHost{
+				Status: bmh.BareMetalHostStatus{
+					HardwareDetails: &bmh.HardwareDetails{
+						NIC: []bmh.NIC{nic1},
+					},
+				},
+			},
+			ExpectedNodeAddresses: []capi.MachineAddress{addr1},
+		},
+		"Two NICs": {
+			Host: &bmh.BareMetalHost{
+				Status: bmh.BareMetalHostStatus{
+					HardwareDetails: &bmh.HardwareDetails{
+						NIC: []bmh.NIC{nic1, nic2},
+					},
+				},
+			},
+			ExpectedNodeAddresses: []capi.MachineAddress{addr1, addr2},
+		},
+		"Hostname is set": {
+			Host: &bmh.BareMetalHost{
+				Status: bmh.BareMetalHostStatus{
+					HardwareDetails: &bmh.HardwareDetails{
+						Hostname: "mygreathost",
+					},
+				},
+			},
+			ExpectedNodeAddresses: []capi.MachineAddress{addr3},
+		},
+		"Empty Hostname": {
+			Host: &bmh.BareMetalHost{
+				Status: bmh.BareMetalHostStatus{
+					HardwareDetails: &bmh.HardwareDetails{
+						Hostname: "",
+					},
+				},
+			},
+			ExpectedNodeAddresses: []capi.MachineAddress{},
+		},
+		"No host at all, so this is a no-op": {
+			Host:                  nil,
+			ExpectedNodeAddresses: nil,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Logf("## TC-%s ##", name)
+		var nodeAddresses []capi.MachineAddress
+
+		c := fakeclient.NewFakeClientWithScheme(scheme)
+		machineMgr, err := newMachineManager(c, nil, nil, &tc.Machine, &tc.BMMachine, klogr.New())
+
+		if err != nil {
+			t.Error(err)
+		}
+
+		if tc.Host != nil {
+			nodeAddresses = machineMgr.nodeAddresses(tc.Host)
+			if err != nil {
+				t.Errorf("unexpected error %v", err)
+			}
+		}
+		for i, address := range tc.ExpectedNodeAddresses {
+			if address != nodeAddresses[i] {
+				t.Errorf("expected Address %v, found %v", address, nodeAddresses[i])
+			}
+		}
+	}
+}
+
+func newConfig(t *testing.T,
+	UserDataNamespace string,
+	labels map[string]string,
+	reqs []capbm.HostSelectorRequirement) (*capbm.BareMetalMachine, *corev1.ObjectReference) {
+	config := capbm.BareMetalMachine{
+		Spec: capbm.BareMetalMachineSpec{
+			Image: capbm.Image{
+				URL:      testImageURL,
+				Checksum: testImageChecksumURL,
+			},
+			UserData: &corev1.SecretReference{
+				Name:      testUserDataSecretName,
+				Namespace: UserDataNamespace,
+			},
+			HostSelector: capbm.HostSelector{
+				MatchLabels:      labels,
+				MatchExpressions: reqs,
+			},
+		},
+	}
+
+	infrastructureRef := &corev1.ObjectReference{
+		Name:       "someothermachine",
+		Namespace:  "myns",
+		Kind:       "Machine",
+		APIVersion: capi.GroupVersion.String(),
+	}
+	return &config, infrastructureRef
+}
+
+func newMachine(machineName string, bareMetalMachineName string, hostAnnotation string) *capi.Machine {
```

Comment:
Those three functions could maybe be merged with what is in the test suite setup.